test/integration/acoustic_comms.cc

```diff
@@ -0,0 +1,346 @@
+
+  // Run server
+  size_t iters = 1000;
+  server.Run(true, iters, false);
+
+  unsigned int msgCounter = 0u;
+  std::mutex mutex;
+  auto cb = [&](const msgs::Dataframe &_msg) -> void
+  {
+    // Verify msg content
+    std::lock_guard<std::mutex> lock(mutex);
+    std::string expected = "hello world " + std::to_string(msgCounter);
+
+    gz::msgs::StringMsg receivedMsg;
+    receivedMsg.ParseFromString(_msg.data());
+    EXPECT_EQ(expected, receivedMsg.data());
+    msgCounter++;
+  };
+
+  // Create subscriber.
+  gz::transport::Node node;
+  std::string addr  = "addr1";
+  std::string subscriptionTopic = "addr1/rx";
+
+  // Subscribe to a topic by registering a callback.
+  auto cbFunc = std::function<void(const msgs::Dataframe &)>(cb);
+  EXPECT_TRUE(node.Subscribe(subscriptionTopic, cbFunc))
+      << "Error subscribing to topic [" << subscriptionTopic << "]";
+
+  // Create publisher.
+  std::string publicationTopic = "/broker/msgs";
+  auto pub = node.Advertise<gz::msgs::Dataframe>(publicationTopic);
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  // Prepare the message.
+  gz::msgs::Dataframe msg;
+  msg.set_src_address("addr2");
+  msg.set_dst_address(addr);
+
+  // Publish some messages.
+  gz::msgs::StringMsg payload;
+  unsigned int pubCount = 3u;
+  for (unsigned int i = 0u; i < pubCount; ++i)
+  {
+    // Prepare the payload.
+    payload.set_data("hello world " + std::to_string(i));
+    std::string serializedData;
+    EXPECT_TRUE(payload.SerializeToString(&serializedData))
+        << payload.DebugString();
+    msg.set_data(serializedData);
+    EXPECT_TRUE(pub.Publish(msg));
+    server.Run(true, 100, false);
+  }
+
+  // Verify subscriber received all msgs.
+  int sleep = 0;
+  bool done = false;
+  while (!done && sleep++ < 3)
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    done = msgCounter == pubCount;
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+  EXPECT_EQ(pubCount, msgCounter);
+}
+
+/////////////////////////////////////////////////
+TEST_F(AcousticCommsTest,
+    GZ_UTILS_TEST_DISABLED_ON_WIN32(AcousticCommsMovingDst))
+{
+  // Start server
+  ServerConfig serverConfig;
+  const auto sdfFile =
+    gz::common::joinPaths(std::string(PROJECT_SOURCE_PATH),
+      "examples", "worlds", "acoustic_comms_moving_targets.sdf");
+  serverConfig.SetSdfFile(sdfFile);
+
+  Server server(serverConfig);
+  EXPECT_FALSE(server.Running());
+  EXPECT_FALSE(*server.Running(0));
+
+  // Run server
+  size_t iters = 1000;
+  server.Run(true, iters, false);
+
+  unsigned int msgCounter = 0u;
+  std::mutex mutex;
+  auto cb = [&](const msgs::Dataframe &_msg) -> void
+  {
+    // Verify msg content
+    std::lock_guard<std::mutex> lock(mutex);
+    std::string expected = "hello world " + std::to_string(msgCounter);
+
+    gz::msgs::StringMsg receivedMsg;
+    receivedMsg.ParseFromString(_msg.data());
+    EXPECT_EQ(expected, receivedMsg.data());
+    msgCounter++;
+  };
```

Comment:
@adityapande-1995 meta: I see lots of duplication, can we use a fixture?

Reply:
Done https://github.com/gazebosim/gz-sim/pull/1608/commits/8f643c5cff0393961e2c7e68d5f4b1792a193efb